src/main/java/com/ably/kafka/connect/client/DefaultAblyBatchClient.java

```diff
@@ -0,0 +1,154 @@
+package com.ably.kafka.connect.client;
+
+import com.ably.kafka.connect.config.ChannelSinkConnectorConfig;
+import static com.ably.kafka.connect.config.ChannelSinkConnectorConfig.*;
+import com.ably.kafka.connect.config.ConfigValueEvaluator;
+import com.ably.kafka.connect.mapping.ChannelSinkMapping;
+import com.ably.kafka.connect.mapping.MessageSinkMapping;
+import io.ably.lib.http.HttpCore;
+import io.ably.lib.http.HttpUtils;
+import io.ably.lib.rest.AblyRest;
+import io.ably.lib.types.AblyException;
+import io.ably.lib.types.HttpPaginatedResponse;
+import io.ably.lib.types.Message;
+import io.ably.lib.types.Param;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+
+/**
+ * Ably Batch client based on REST API
+ */
+public class DefaultAblyBatchClient implements AblyClient {
+    private static final Logger logger = LoggerFactory.getLogger(DefaultAblyBatchClient.class);
+
+    protected final ChannelSinkMapping channelSinkMapping;
+    protected final MessageSinkMapping messageSinkMapping;
+    protected final ChannelSinkConnectorConfig connectorConfig;
+
+    private final ConfigValueEvaluator configValueEvaluator;
+    AblyRest restClient;
+
+    public DefaultAblyBatchClient(ChannelSinkConnectorConfig connectorConfig, ChannelSinkMapping channelSinkMapping,
+                                  MessageSinkMapping messageSinkMapping, ConfigValueEvaluator configValueEvaluator) {
+        this.connectorConfig = connectorConfig;
+        this.channelSinkMapping = channelSinkMapping;
+        this.messageSinkMapping = messageSinkMapping;
+        this.configValueEvaluator = configValueEvaluator;
+
+    }
+
+    @Override
+    public void connect() throws ConnectException, AblyException {
+        this.restClient = new AblyRest(connectorConfig.getPassword(CLIENT_KEY).value());
+    }
+
+    /**
+     * Function that uses the REST API client to send batches.
+     * @param records list of sink records.
+     * @throws ConnectException
+     */
+    public void publishBatch(List<SinkRecord> records) throws ConnectException, AblyException {
+
+        if(!records.isEmpty()) {
+            Map<String, List<Message>> groupedMessages = groupMessagesByChannel(records);
+
+            if(groupedMessages != null && !groupedMessages.isEmpty())
+            {
+                groupedMessages.forEach((key, value) -> {
+                    logger.debug("Ably BATCH call -Thread(" + Thread.currentThread().getName() + ")" + "Num Records:" + value.size());
+                    try {
+                        sendBatches(new BatchSpec(Set.of(key), value));
```

Comment:
Am I reading this correctly?

The publishBatch call will:

* group messages by channel
* iterate through each channel 
* publish messages for a single channel as one batch
* given publish is sync, all publishing will be stacked up based on the number of channels, so if we have messages grouped into 100 channels, we'd have 100 serial requests made to the batch API, so assuming a 150ms roundtrip, tat would take 15 seconds.

If this is the case, I think this is wrong.

The goal, as defined in the document & Gist we shared is to:

* Construct batches based on the known constraints 
* Publish those batches based on the max concurrency available i.e. if there are 10 threads, and 20 batches, we'd publish 10, wait for one of those threads to free up, and then publish the next batch on the queue.

Reply:
@mattheworiordan grouping was done based on this comment - https://github.com/ably/kafka-connect-ably/pull/114#discussion_r1185109116 
But it doesn't talk about threads.

Reply:
Please check updated logic of using thread pool to call REST API in parallel.

Reply:
```
connector          | [2023-05-21 19:02:12,864] INFO Ably BATCH call -Thread(pool-6-thread-2) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:12,864] INFO Ably BATCH call -Thread(pool-6-thread-1) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:13,238] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:13,612] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:17,865] INFO Ably BATCH call -Thread(pool-6-thread-4) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:17,865] INFO Ably BATCH call -Thread(pool-6-thread-3) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:18,441] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:18,691] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:22,865] INFO Ably BATCH call -Thread(pool-6-thread-6) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:22,865] INFO Ably BATCH call -Thread(pool-6-thread-5) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:23,426] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:23,504] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:27,866] INFO Ably BATCH call -Thread(pool-6-thread-7) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:27,866] INFO Ably BATCH call -Thread(pool-6-thread-8) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:28,204] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:28,537] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:32,865] INFO Ably BATCH call -Thread(pool-6-thread-10) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:32,865] INFO Ably BATCH call -Thread(pool-6-thread-9) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:33,426] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:33,713] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:37,866] INFO Ably BATCH call -Thread(pool-6-thread-1) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:37,866] INFO Ably BATCH call -Thread(pool-6-thread-2) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:38,458] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:38,623] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:42,865] INFO Ably BATCH call -Thread(pool-6-thread-4) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:42,865] INFO Ably BATCH call -Thread(pool-6-thread-3) (com.ably.kafka.connect.batch.BatchProcessingThread:29)
connector          | [2023-05-21 19:02:43,275] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
connector          | [2023-05-21 19:02:43,353] INFO Response: 201 error: 0 - null (com.ably.kafka.connect.client.DefaultAblyBatchClient:142)
```

Reply:
I can see from this log that all 10 threads are used, but it's odd that there's a fixed level of concurrency - ie consistently there are only two requests in flight. Is there something about your test (eg the rate of adding SinkRecords) that would explain that?

Reply:
From comment 114 linked above:

> You then construct a BatchSpec per outgoing channel name and **submit them as a list.**

The last part is what was missing -- you can submit a *list* of `BatchSpec` instances using the Batch API, which isn't super obvious but was demonstrated in the original gist, [here](https://gist.github.com/jaley/a32cf6b9974af4ce68a029cc601d8cd0#file-ablybatchv2example-java-L38). The implementation of `sendBatches()` here keeps the plural in the function name but only accepts a single `BatchSpec` as a parameter.

Reply:
I think perhaps the behaviour @paddybyers mentions above, where we see a max of two active threads may be related to the race I describe above. The level of parallelism realised in practice with this design will be quite hard to predict but it'll be some complicated function of:

* Time taken to submit each batch request
* Scheduling of `SinkRecordsProcessingThread` against the main `ChannelSinkTask` thread - i.e. how often the former polls an empty queue because the latter hasn't run for a while
* The polling interval configuration and rate of incoming data

So it makes sense that you can see work being scheduled to every thread in the pool, but the number active at any given time being much smaller than the pool size. We should discuss the parallelism strategy on a call to explore higher throughput options